src/Workspaces/Core/MSBuild/MSBuild/MSBuildWorkspace_ProjectReload.cs

```diff
@@ -0,0 +1,309 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Serialization;
+using Microsoft.CodeAnalysis.Shared.Collections;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.MSBuild
+{
+    public sealed partial class MSBuildWorkspace
+    {
+        private async Task<ImmutableArray<Project>> UpdateProjectsAsync(ImmutableArray<ProjectInfo> newProjectInfos)
+        {
+            using (_serializationLock.DisposableWait())
+            {
+                var serialization = this.Services.GetRequiredService<ISerializerService>();
+                var updatedProjectIds = TemporaryArray<ProjectId>.Empty;
+                var oldSolution = this.CurrentSolution;
+                var solution = oldSolution;
+                foreach (var projectInfo in newProjectInfos)
+                {
+                    var projectToUpdate = solution.GetProject(projectInfo.Id);
+                    RoslynDebug.AssertNotNull(projectToUpdate);
+                    var oldCheckSum = await projectToUpdate.State.GetStateChecksumsAsync(default).ConfigureAwait(false);
+                    var newCheckSum = projectInfo.GetCheckSum(serialization);
+                    var newProject = await UpdateProjectAsync(projectToUpdate, projectInfo, oldCheckSum, newCheckSum).ConfigureAwait(false);
+                    solution = AdjustReloadedProject(projectToUpdate, newProject).Solution;
+                    updatedProjectIds.Add(newProject.Id);
+                }
+
+                this.SetCurrentSolution(solution);
+                using var _1 = ArrayBuilder<Project>.GetInstance(out var updatedProjects);
+                foreach (var projectId in updatedProjectIds)
+                {
+                    _ = this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.ProjectReloaded, oldSolution, solution, projectId);
```

Comment:
Raising multiple times with the old and new isn't something we really expect to happen. I think we did this once around linked files and it was...strange. If nothing else I really don't have confidence the diagnostic engine will handle this well -- should the be raising a single solution change if there are multiple projects being touched?